crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -38,6 +45,209 @@ struct QueryLimitsCheckerExt {
+        available_budget: &mut u64,
+        variables_length: &mut HashMap<String, u64>,
+        items: &[Positioned<Selection>],
+        fragments: &HashMap<Name, Positioned<FragmentDefinition>>,
+    ) -> ServerResult<()> {
+        for n in items {
+            match &n.node {
+                Selection::Field(f) => {
+                    if f.node.name.node == DRY_RUN_TX_BLOCK || f.node.name.node == EXECUTE_TX_BLOCK
+                    {
+                        for arg in &f.node.arguments {
+                            check_node(
+                                arg,
+                                available_budget,
+                                variables_length,
+                                variables,
+                                max_tx_payload_size,
+                            )?;
+                        }
+                    }
+                }
+                Selection::FragmentSpread(fs) => {
+                    let name = &fs.node.fragment_name.node;
+                    let def = fragments.get(name).ok_or_else(|| {
+                        graphql_error_at_pos(
+                            code::INTERNAL_SERVER_ERROR,
+                            format!("Fragment {name} referred to but not found in document"),
+                            fs.pos,
+                        )
+                    })?;
+                    traverse_and_check_tx_payload(
+                        variables,
+                        max_tx_payload_size,
+                        available_budget,
+                        variables_length,
+                        &def.node.selection_set.node.items,
+                        fragments,
+                    )?;
+                }
+                Selection::InlineFragment(f) => {
+                    traverse_and_check_tx_payload(
+                        variables,
+                        max_tx_payload_size,
+                        available_budget,
+                        variables_length,
+                        &f.node.selection_set.node.items,
+                        fragments,
+                    )?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn check_node(
+        node: &(Positioned<Name>, Positioned<Value>),
+        available_budget: &mut u64,
+        variables_length: &mut HashMap<String, u64>,
+        variables: &Variables,
+        max_tx_payload_size: u64,
+    ) -> ServerResult<()> {
+        let node_name = node.0.node.to_string();
+        let node_len = if is_variable(&node.1.node) {
+            if let Some(v) = variables_length.get(&node_name) {
+                *v
+            } else {
+                let node_len = get_value_str_len(&node.1.node, variables).map_err(|_| {
+                    graphql_error_at_pos(
+                        code::INTERNAL_SERVER_ERROR,
+                        format!("Error getting size of variable: {}", node_name),
+                        node.1.pos,
+                    )
```

Comment:
Is it possible that we always discard the error we get from `get_value_str_len` and replace it with this error, and if so, is it worth updating it so that this is the error it produces in the first place? (Also similar to the above, this should probably be a client error, not an internal error).

Reply:
yes, we're not reporting the real error here.

Reply:
Indeed, fixed it.